### PR TITLE
ARO clusters support

### PR DESCRIPTION
**Pull request by omrirh**
Jira ticket https://issues.redhat.com/browse/MSINTEROP-647
Depends on https://github.com/RedHatQE/cloud-tools/pull/475

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -277,6 +289,22 @@ podman run quay.io/redhat_msi/openshift-cli-installer \
     --cluster 'name=hyper;platform=hypershift;region=us-west-2;version=4.13.4;compute-machine-type=m5.4xlarge;replicas=6;channel-group=candidate;expiration-time=4h;timeout=1h'
 ```
 
+##### ARO cluster
+
+###### Versions
+  * Current supported versions for ARO are 4.13.23 and 4.12.25
```
Comment (rnetser) [inline README.md:295]: please add an explanation how to get supported versions and not hard-code them here

File: README.md [hunk 2]
```diff
@@ -326,6 +354,19 @@ podman run quay.io/redhat_msi/openshift-cli-installer \
...
+
+```
+podman run quay.io/redhat_msi/openshift-cli-installer \
+    --action destroy \
+    --ocm-token=$OCM_TOKEN \
+    --azure-tenant-id=$AZURE_TENANT_ID \
+    --azure-client-id=$AZURE_CLIENT_ID \
+    --azure-client-secret=$AZURE_CLIENT_SECRET \
+    --azure-subscription-id=$AZURE_SUBSCRIPTION_ID \
+    --cluster 'name=aro1;platform=aro;region=eastus;version=4.13.23;master-vm-size=Standard_D8s_v3;workers-vm-size=Standard_D4s_v3;workers-count=4;timeout=1h'
```
Comment (rnetser) [inline README.md:367]: please fix in-solidarity

File: openshift_cli_installer/cli.py [hunk 3]
```diff
@@ -199,6 +199,38 @@
...
+\b
+Azure's service principal client secret.
+""",
+    type=str,
+)
+@click.option(
+    "--azure-subscription-id",
+    help="""
+\b
+MSI Azure subscription ID.
```
Comment (rnetser) [inline openshift_cli_installer/cli.py:230]:
```suggestion
Azure subscription ID.
```

File: openshift_cli_installer/libs/clusters/ocp_cluster.py [hunk 4]
```diff
@@ -94,6 +98,7 @@ def __init__(self, ocp_cluster, user_input):
         self.log_prefix = f"[C:{self.cluster_info['name']}|P:{self.cluster_info['platform']}|R:{self.cluster_info.get('region', 'auto-region')}]"
         self.timeout = tts(ts=self.cluster.get("timeout", TIMEOUT_60MIN))
 
+        self._cluster_data_yaml_file = os.path.join(self.cluster_info["cluster-dir"], CLUSTER_DATA_YAML_FILENAME)
```
Comment (rnetser) [inline openshift_cli_installer/libs/clusters/ocp_cluster.py:101]: remove "_"

File: openshift_cli_installer/libs/clusters/aro_cluster.py [hunk 5]
```diff
@@ -0,0 +1,248 @@
...
+from simple_logger.logger import get_logger
+
+
+class AROCluster(OCPCluster):
+    def __init__(self, ocp_cluster, user_input):
+        super().__init__(ocp_cluster=ocp_cluster, user_input=user_input)
+        self.logger = get_logger(f"{self.__class__.__module__}-{self.__class__.__name__}")
+        self.set_azure_authentication()
+
+        if not self.user_input.create:
```
Comment (rnetser) [inline openshift_cli_installer/libs/clusters/aro_cluster.py:22]: self.create

and reverse the logic - positive condition

File: openshift_cli_installer/libs/clusters/aro_cluster.py [hunk 6]
```diff
@@ -0,0 +1,248 @@
...
+            client_secret=self.client_secret,
+        )
+
+        self.aro_client, self.network_client, self.resource_client = (
+            client(credential=self.credentials, subscription_id=self.subscription_id)
+            for client in (aro_client, network_client, resource_client)
+        )
+
+    def set_aro_cluster_params(self, cluster_data):
+        self.cluster_name = self.get_cluster_name()
```
Comment (rnetser) [inline openshift_cli_installer/libs/clusters/aro_cluster.py:46]: stored in cluster-info

File: openshift_cli_installer/utils/general.py [hunk 7]
```diff
@@ -114,3 +116,7 @@ def get_local_ssh_key(ssh_key_file):
 def get_dict_from_json(gcp_service_account_file):
     with open(gcp_service_account_file) as fd:
         return json.loads(fd.read())
+
+
+def random_resource_postfix(length=4):
```
Comment (rnetser) [inline openshift_cli_installer/utils/general.py:121]: please re-use openshift_cli_installer.libs.clusters.rosa_cluster.RosaCluster.generate_hypershift_password

File: openshift_cli_installer/libs/clusters/aro_cluster.py [hunk 8]
```diff
@@ -0,0 +1,248 @@
...
+        cluster_resources_postfix = random_resource_postfix()
+        resource_group_name_str = "resource-group-name"
+        virtual_network_name_str = "virtual-network-name"
+
+        aro_cluster_args = {
+            "domain": "msi",
```
Comment (rnetser) [inline openshift_cli_installer/libs/clusters/aro_cluster.py:52]: do not hard-code the domain

File: openshift_cli_installer/libs/clusters/aro_cluster.py [hunk 9]
```diff
@@ -0,0 +1,248 @@
...
+            resource_group_name_str: f"aro-rg-{cluster_resources_postfix}",
+            "cluster-resource-group-name": f"{self.cluster_name}-rg-{cluster_resources_postfix}",
+            virtual_network_name_str: f"aro-vnet-{cluster_resources_postfix}",
+            "workers-subnet-name": f"workers-subnet-{cluster_resources_postfix}",
+            "master-subnet-name": f"master-subnet-{cluster_resources_postfix}",
+            "master-vm-size": "Standard_D8s_v3",
```
Comment (rnetser) [inline openshift_cli_installer/libs/clusters/aro_cluster.py:58]: get it from the user, you can default to something

File: openshift_cli_installer/libs/clusters/aro_cluster.py [hunk 10]
```diff
@@ -0,0 +1,248 @@
...
+            "workers-vm-size": "Standard_D4s_v3",
+            "workers-count": 3,
+            "workers-disk-size": 128,
+            "network-pod-cidr": "10.128.0.0/14",
+            "network-service-cidr": "172.30.0.0/16",
+            "fips": False,
```
Comment (rnetser) [inline openshift_cli_installer/libs/clusters/aro_cluster.py:64]: ditto

File: openshift_cli_installer/libs/clusters/aro_cluster.py [hunk 11]
```diff
@@ -0,0 +1,248 @@
...
+        }
+        self.cluster.update({k: cluster_data.pop(k, v) for k, v in aro_cluster_args.items()})
```
Comment (rnetser) [inline openshift_cli_installer/libs/clusters/aro_cluster.py:66]: use meaningful variable names

File: openshift_cli_installer/libs/clusters/aro_cluster.py [hunk 12]
```diff
@@ -0,0 +1,248 @@
...
+
+        if pull_secret_file := self.user_input.docker_config_file:
```
Comment (rnetser) [inline openshift_cli_installer/libs/clusters/aro_cluster.py:68]: this is not mandatory?

File: openshift_cli_installer/libs/clusters/aro_cluster.py [hunk 13]
```diff
@@ -0,0 +1,248 @@
...
+            with open(pull_secret_file, "r") as ps_file:
+                self.pull_secret = ps_file.read()
+
+        self.region = self.cluster["region"]
+        self.version = self.cluster["version"]
+        self.platform = self.cluster_info["platform"]
```
Comment (rnetser) [inline openshift_cli_installer/libs/clusters/aro_cluster.py:74]: consume directly where needed, no need to assign

File: openshift_cli_installer/libs/clusters/aro_cluster.py [hunk 14]
```diff
@@ -0,0 +1,248 @@
...
+        self.resource_group_name = self.cluster[resource_group_name_str]
+        self.virtual_network_name = self.cluster[virtual_network_name_str]
+
+        self.assert_cluster_params_are_valid()
```
Comment (rnetser) [inline openshift_cli_installer/libs/clusters/aro_cluster.py:78]: should be in `user_input`

File: openshift_cli_installer/libs/clusters/aro_cluster.py [hunk 15]
```diff
@@ -0,0 +1,248 @@
...
+        if not self.user_input.destroy_from_s3_bucket_or_local_directory:
+            self.dump_aro_cluster_data_to_file()
+
+    def dump_aro_cluster_data_to_file(self):
```
Comment (rnetser) [inline openshift_cli_installer/libs/clusters/aro_cluster.py:82]: why do we need a separate function for this?

File: openshift_cli_installer/libs/clusters/aro_cluster.py [hunk 16]
```diff
@@ -0,0 +1,248 @@
...
+            resource_group_name=self.resource_group_name, parameters={"location": self.region}
+        )
+
+    def delete_resource_group(self):
+        self.logger.info(f"Deleting resource group {self.resource_group_name}")
+        self.resource_client.resource_groups.begin_delete(resource_group_name=self.resource_group_name)
+
+    def create_virtual_network(
+        self,
+        vnet_address_prefix="10.0.0.0/22",
```
Comment (rnetser) [inline openshift_cli_installer/libs/clusters/aro_cluster.py:122]: get from user with defaults

File: openshift_cli_installer/libs/clusters/aro_cluster.py [hunk 17]
```diff
@@ -0,0 +1,248 @@
...
+        self.network_client.virtual_networks.begin_create_or_update(
+            virtual_network_name=self.virtual_network_name,
+            resource_group_name=self.resource_group_name,
+            parameters={
+                "location": self.region,
+                "address_space": {"address_prefixes": [vnet_address_prefix]},
+            },
+        ).result()
+
+        # Create worker + master nodes subnets
```
Comment (rnetser) [inline openshift_cli_installer/libs/clusters/aro_cluster.py:145]: please remove the comment; code should be explicit

File: openshift_cli_installer/libs/clusters/aro_cluster.py [hunk 18]
```diff
@@ -0,0 +1,248 @@
...
+        )
+
+    def delete_virtual_network(self):
+        self.logger.info(f"Deleting virtual network {self.virtual_network_name}")
+        self.network_client.virtual_networks.begin_delete(
+            virtual_network_name=self.virtual_network_name,
+            resource_group_name=self.resource_group_name,
+        ).result()
+
+    def aro_cluster_kubeconfig(self):
```
Comment (rnetser) [inline openshift_cli_installer/libs/clusters/aro_cluster.py:160]: should be in ocm wrapper

File: openshift_cli_installer/libs/clusters/aro_cluster.py [hunk 19]
```diff
@@ -0,0 +1,248 @@
...
+        self.logger.info(f"Fetching {self.platform} cluster {self.cluster_name} kubeconfig")
+        return (
+            base64.b64decode(
+                s=self.aro_client.open_shift_clusters.list_admin_credentials(
+                    resource_group_name=self.resource_group_name, resource_name=self.cluster_name
+                ).kubeconfig
+            )
+        ).decode("utf-8")
+
+    def aro_cluster_kubeadmin_password(self):
```
Comment (rnetser) [inline openshift_cli_installer/libs/clusters/aro_cluster.py:170]: ditto

File: openshift_cli_installer/libs/clusters/aro_cluster.py [hunk 20]
```diff
@@ -0,0 +1,248 @@
...
+        self.logger.info(f"Fetching {self.platform} cluster {self.cluster_name} kubeadmin-password")
+        return self.aro_client.open_shift_clusters.list_credentials(
+            resource_group_name=self.resource_group_name,
+            resource_name=self.cluster_name,
+        ).kubeadmin_password
+
+    def set_aro_cluster_auth(self):
```
Comment (rnetser) [inline openshift_cli_installer/libs/clusters/aro_cluster.py:177]: already exists

File: openshift_cli_installer/libs/clusters/aro_cluster.py [hunk 21]
```diff
@@ -0,0 +1,248 @@
...
+
+    def create_cluster(self):
+        self.create_cluster_resources()
+        self.timeout_watch = self.start_time_watcher()
+
+        cluster_body = {
+            "clusterProfile": {
+                "domain": self.cluster["domain"],
+                "fipsValidatedModules": "Enabled" if self.cluster["fips"] else "Disabled",
+                "pullSecret": self.pull_secret,
```
Comment (rnetser) [inline openshift_cli_installer/libs/clusters/aro_cluster.py:193]: can this be None?

File: openshift_cli_installer/libs/clusters/aro_cluster.py [hunk 22]
```diff
@@ -0,0 +1,248 @@
...
+        auth_path = self.cluster_info["auth-path"]
+        with open(os.path.join(auth_path, "kubeconfig"), "w") as kubeconfig:
+            kubeconfig.write(self.aro_cluster_kubeconfig())
+
+        with open(os.path.join(auth_path, "kubeadmin-password"), "w") as kubeadmin:
+            kubeadmin.write(self.aro_cluster_kubeadmin_password())
+
+    def create_cluster(self):
+        self.create_cluster_resources()
+        self.timeout_watch = self.start_time_watcher()
```
Comment (rnetser) [inline openshift_cli_installer/libs/clusters/aro_cluster.py:187]: why is this needed?

File: openshift_cli_installer/libs/clusters/aro_cluster.py [hunk 23]
```diff
@@ -0,0 +1,248 @@
...
+            resource_name=self.cluster_name,
+            resource_group_name=self.resource_group_name,
+            parameters={"location": self.region, "properties": cluster_body},
+        ).result()
+        self.set_aro_cluster_auth()
+
+        self.logger.info(f"{self.platform} cluster {self.cluster_name} created successfully.")
+
+    def destroy_cluster(self):
+        self.timeout_watch = self.start_time_watcher()
```
Comment (rnetser) [inline openshift_cli_installer/libs/clusters/aro_cluster.py:239]: why needed?

File: openshift_cli_installer/libs/clusters/aro_cluster.py [hunk 24]
```diff
@@ -0,0 +1,248 @@
...
+
+        self.logger.info(f"Creating {self.platform} cluster {self.cluster_name}")
+        self.aro_client.open_shift_clusters.begin_create_or_update(
+            resource_name=self.cluster_name,
+            resource_group_name=self.resource_group_name,
+            parameters={"location": self.region, "properties": cluster_body},
+        ).result()
+        self.set_aro_cluster_auth()
+
+        self.logger.info(f"{self.platform} cluster {self.cluster_name} created successfully.")
```
Comment (rnetser) [inline openshift_cli_installer/libs/clusters/aro_cluster.py:236]: if fail; we need to rollback

File: openshift_cli_installer/libs/clusters/ocp_cluster.py [hunk 25]
```diff
@@ -48,8 +48,12 @@ def __init__(self, ocp_cluster, user_input):
         self.logger = get_logger(f"{self.__class__.__module__}-{self.__class__.__name__}")
         self.cluster = ocp_cluster
 
-        self.s3_bucket_name = self.user_input.s3_bucket_name or self.cluster["cluster_info"].get("s3_bucket_name")
-        self.s3_bucket_path = self.user_input.s3_bucket_path or self.cluster["cluster_info"].get("s3_bucket_path")
+        self.s3_bucket_name = self.user_input.s3_bucket_name or (
+            self.cluster["cluster_info"].get("s3_bucket_name") if self.cluster.get("cluster_info") else None
```
Comment (rnetser) [inline openshift_cli_installer/libs/clusters/ocp_cluster.py:52]: why is this change needed?

File: openshift_cli_installer/manifests/clusters.example.yaml [hunk 26]
```diff
@@ -144,3 +144,23 @@ clusters:
   worker-flavor: m5.4xlarge
   worker-root-disk-size: 128
   log_level: info # optional, default: "error", supported options are debug, info, warn, error
+
+# ARO cluster
+- name: msi-aro-c1
```
Comment (rnetser) [inline openshift_cli_installer/manifests/clusters.example.yaml:149]: please remove `msi` from examples and code

File: openshift_cli_installer/manifests/clusters.example.yaml [hunk 27]
```diff
@@ -144,3 +144,23 @@ clusters:
...
+  platform: aro
+  version: "4.13.23"
+  region: eastus
+  stream: stable
+  timeout: 1h
+  domain: msi # Domain name must contain 1 to 30 alphanumeric characters or '.', and start and end with an alphabetic character
```
Comment (rnetser) [inline openshift_cli_installer/manifests/clusters.example.yaml:155]: the comment is about valid domains; please remove.
add to readme that domain must be a valid one

**Review by rnetser — CHANGES_REQUESTED** [review]

File: openshift_cli_installer/libs/clusters/aro_cluster.py (hunk 23 above)
Comment (omrirh) [reply openshift_cli_installer/libs/clusters/aro_cluster.py:239]: will remove

File: openshift_cli_installer/libs/clusters/ocp_cluster.py (hunk 25 above)
Comment (omrirh) [reply openshift_cli_installer/libs/clusters/ocp_cluster.py:52]: `self.cluster["cluster_info"]` does not exist on cluster create.
It gets written to cluster data yaml file on init (`cluster_info` not in [keys_to_pop](https://github.com/RedHatQE/openshift-cli-installer/blob/dfabe4706b7cbd891d3cc73bd8b52943e772f66f/openshift_cli_installer/libs/clusters/ocp_cluster.py#L189))

File: openshift_cli_installer/libs/clusters/aro_cluster.py (hunk 9 above)
Comment (omrirh) [reply openshift_cli_installer/libs/clusters/aro_cluster.py:58]: It is being retrieved from the user if exists via `self.cluster_info`, and uses this default value if not specified from this dict.
`self.cluster.update({k: cluster_data.pop(k, v) for k, v in aro_cluster_args.items()})`  will pop the specified keys from `self.cluster_info` (`cluster_data=self.cluster_info` on create)

File: openshift_cli_installer/libs/clusters/aro_cluster.py (hunk 10 above)
Comment (omrirh) [reply openshift_cli_installer/libs/clusters/aro_cluster.py:64]: [ditto](https://github.com/RedHatQE/openshift-cli-installer/pull/602#discussion_r1555554906)

File: openshift_cli_installer/libs/clusters/aro_cluster.py (hunk 5 above)
Comment (omrirh) [reply openshift_cli_installer/libs/clusters/aro_cluster.py:22]: only `user_input` has the create attribute.
I can add `self.create = self.user_input.create` on `ocp_cluster.py` init if needed, let me know

File: openshift_cli_installer/cli.py [hunk 28]
```diff
@@ -199,6 +199,38 @@
...
+@click.option(
+    "--azure-client-id",
+    help="""
+\b
+Azure's service principal client ID.
+""",
+    type=str,
+)
+@click.option(
+    "--azure-client-secret",
```
Comment (chetna14manku) [inline openshift_cli_installer/cli.py:219]: Can we get default values for AZURE taken from `os.environ`, like we do for AWS values?

File: openshift_cli_installer/libs/clusters/aro_cluster.py (hunk 15 above)
Comment (omrirh) [reply openshift_cli_installer/libs/clusters/aro_cluster.py:82]: Will re-call `dump_cluster_data_to_file` instead

File: openshift_cli_installer/libs/clusters/aro_cluster.py (hunk 22 above)
Comment (omrirh) [reply openshift_cli_installer/libs/clusters/aro_cluster.py:187]: will remove

File: README.md (hunk 1 above)
Comment (omrirh) [reply README.md]: done

File: openshift_cli_installer/libs/clusters/ocp_cluster.py (hunk 4 above)
Comment (omrirh) [reply openshift_cli_installer/libs/clusters/ocp_cluster.py]: done

File: openshift_cli_installer/libs/clusters/aro_cluster.py (hunk 6 above)
Comment (omrirh) [reply openshift_cli_installer/libs/clusters/aro_cluster.py]: done

File: openshift_cli_installer/utils/general.py (hunk 7 above)
Comment (omrirh) [reply openshift_cli_installer/utils/general.py:121]: done

File: openshift_cli_installer/libs/clusters/aro_cluster.py (hunk 8 above)
Comment (omrirh) [reply openshift_cli_installer/libs/clusters/aro_cluster.py]: done

File: openshift_cli_installer/libs/clusters/aro_cluster.py (hunk 11 above)
Comment (omrirh) [reply openshift_cli_installer/libs/clusters/aro_cluster.py]: done

File: openshift_cli_installer/libs/clusters/aro_cluster.py (hunk 13 above)
Comment (omrirh) [reply openshift_cli_installer/libs/clusters/aro_cluster.py]: done

File: openshift_cli_installer/libs/clusters/aro_cluster.py (hunk 16 above)
Comment (omrirh) [reply openshift_cli_installer/libs/clusters/aro_cluster.py]: done

File: openshift_cli_installer/libs/clusters/aro_cluster.py (hunk 17 above)
Comment (omrirh) [reply openshift_cli_installer/libs/clusters/aro_cluster.py]: done

File: openshift_cli_installer/libs/clusters/aro_cluster.py (hunk 22 above)
Comment (omrirh) [reply openshift_cli_installer/libs/clusters/aro_cluster.py]: done

File: openshift_cli_installer/libs/clusters/aro_cluster.py (hunk 23 above)
Comment (omrirh) [reply openshift_cli_installer/libs/clusters/aro_cluster.py]: done

File: openshift_cli_installer/libs/clusters/aro_cluster.py (hunk 24 above)
Comment (omrirh) [reply openshift_cli_installer/libs/clusters/aro_cluster.py]: done, deleting cluster + resources on failure

File: openshift_cli_installer/manifests/clusters.example.yaml (hunk 27 above)
Comment (omrirh) [reply openshift_cli_installer/manifests/clusters.example.yaml]: done

File: openshift_cli_installer/cli.py (hunk 28 above)
Comment (omrirh) [reply openshift_cli_installer/cli.py:221]: done

**Comment by omrirh** [comment]
@rnetser @myakove closing this PR for now as it is in low priority.

**Comment by rnetser** [comment]
@omrirh please close this pr

**Review by rnetser — CHANGES_REQUESTED** [review]
@omrirh please close this pr
